Added CollectD JSON receiver

owais opened the pull request:
Ported CollectD receiver from SignalFX Gateway to OpenTelemetry Collector

Ported from: https://github.com/signalfx/gateway/tree/master/protocol/collectd

tigrannajaryan commented on receiver/collectdreceiver/collectd_test.go:
```diff
@@ -0,0 +1,387 @@
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
```
Unneeded blank line.

tigrannajaryan commented on receiver/collectdreceiver/factory.go:
```diff
@@ -0,0 +1,92 @@
+	"time"
+
+	"go.uber.org/zap"
+
```
Blank line.

tigrannajaryan commented on receiver/collectdreceiver/factory.go:
```diff
@@ -0,0 +1,92 @@
+) (receiver.MetricsReceiver, error) {
+	rCfg := cfg.(*Config)
+	// CollectD receiver only support JSON encoding. We expose a config option
+	// to make it explicit and obvious to the users.
```
👍

tigrannajaryan reviewed (commented):
Please rebase.

owais commented:
@mdubbyap @keitwb can you please take a look as well? Thanks

keitwb commented:
This is really a receiver for the collectd write_http plugin with the `Format "JSON"` option, that should probably be clarified in the docs for this.

Also this is a super-set of the protocol that supports arbitrary dimensions via the square bracket-encoded key value pairs appended to various fields of the value, which should also be clarified in the doc for this receiver as it is technically not fully backwards compatible with the original protocol since you could legally have square brackets in the value fields that could result in unintended conversions for people that weren't using the SignalFx-specific form of collectd.

owais commented:
Updated. @pjanotti please take a look as well when you can.

pjanotti commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+				if cindex < prev {
+					cindex = len(dimensions)
+				}
+				piece := dimensions[prev:cindex]
```
nit: suggestion naming `kv` or `kvPair` instead of `piece`

owais replied:
I guess since it is not guaranteed to be a kvPair at this point, probably best not to call it that?

pjanotti commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+
+			addIfNotNullOrEmpty(labels, "plugin", true, r.Plugin)
+			parseAndAddLabels(labels, r.PluginInstance, r.Host)
+			addIfNotNullOrEmpty(labels, "dsname", !usedDsName, dsName)
```
This is the only place that the 3rd parameter is not the constant true, it seems more readable to have the conditional here and remove the `cond` parameter

pjanotti commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+			parseAndAddLabels(labels, r.PluginInstance, r.Host)
+			addIfNotNullOrEmpty(labels, "dsname", !usedDsName, dsName)
+
+			metric, _ := r.newMetric(metricName, dsType, val, labels)
```
Log as warning? If noise and expected at debug level.

owais replied:
Missed it. Should have handled the error even though it should never happen.

pjanotti commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+			addIfNotNullOrEmpty(labels, "dsname", !usedDsName, dsName)
+
+			metric, _ := r.newMetric(metricName, dsType, val, labels)
+			metrics = append(metrics, metric)
```
Even in case of error? Or errors are actually not possible?

pjanotti commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+	metric := &metricspb.Metric{}
+	point, isDouble, err := r.newPoint(val)
+	if err != nil {
+		return metric, err
```
suggestion: add the name of the metric to the error

pjanotti commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+	lKeys, lValues := labelKeysAndValues(labels)
+	metric.MetricDescriptor = &metricspb.MetricDescriptor{
+		Name:      name,
+		Type:      r.metricType(dsType, isDouble),
```
Comment: this bothers me on the proto: the point has the value but the metric descriptor too... so the format itself allow inconsistencies... am I missing something?

owais replied:
It does. In the sense that descriptor can say the metric is let's say commulative double but each point can be then be either int or double.

tigrannajaryan replied:
No, not improved. Otel metrics still has a type enum in the decsriptor and then a set of fields to store data points. It is possible to misuse it the same way we have here.

pjanotti replied:
Is it worth opening an issue on spec repo? Or there are reasons for that?

pjanotti commented on receiver/collectdreceiver/testdata/config.yaml:
```diff
@@ -0,0 +1,20 @@
+receivers:
+  collectd:
+  collectd/one:
```
nit: add comments here to the configuration options so users can refer/change it to get started.

pjanotti reviewed (commented):
LGTM in general, a bunch of Qs and nits.

owais commented on receiver/collectdreceiver/receiver.go:
```diff
@@ -0,0 +1,182 @@
+		}
+	}
+
+	w.Write([]byte("OK"))
```
This implicitly sets StatusOK.

owais commented on receiver/collectdreceiver/receiver.go:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	w.Write([]byte("OK"))
+	cdr.nextConsumer.ConsumeMetricsData(ctx, md)
```
Good point

owais commented on receiver/collectdreceiver/receiver.go:
```diff
@@ -0,0 +1,182 @@
+	attrs := make(map[string]string)
+	for key := range params {
+		if strings.HasPrefix(key, cdr.defaultAttrsPrefix) {
+			value := params.Get(key)
```
Query() returns Values which is a `map[string][]string` but also has a `Get() string` method which returns the first value for a param. That's what we need.

pjanotti replied:
I see: the lookup feels wateful but `Get` does treatment to handle special cases.

owais commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+// becomes a dimension.
+func (r *collectDRecord) getReasonableMetricName(index int, attrs map[string]string) (string, bool) {
+	usedDsName := false
+	parts := make([]byte, 0, len(*r.TypeS)+len(*r.TypeInstance))
```
Judging from this existing in gateway, I'd say yes but added checks anyway. Thanks

owais commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+				}
+				piece := dimensions[prev:cindex]
+				tindex := strings.Index(piece, "=")
+				//lint:ignore S1003 staticcheck wants us to use strings.Conains instead of Index
```
We use the value of tindex later in the function and on the next line as well so we need to use Index. I think staticcheck wrongly assumes we are only checking for existence of the char at index because `if index == -1` following the .Index() call.

pjanotti replied:
I was think that the `lint:ignore` was for the one on the if statement for which the code just compares the return value.

owais commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+		parts = append(parts, *r.TypeS...)
+	}
+	parts = r.pointTypeInstance(attrs, parts)
+	if r.Dsnames != nil && !isNilOrEmpty(r.Dsnames[index]) && len(r.Dsnames) > 1 {
```
The condition is for kind of a special case when we have more than one dsnames. The function comment covers it I think.

owais commented on receiver/collectdreceiver/receiver.go:
```diff
@@ -0,0 +1,182 @@
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		recordRequestErrors()
+		w.WriteHeader(http.StatusInternalServerError)
```
I thought there might be reasons other than client sending bad data for this but changed to badrequest after reading source of net/http. Most cases should be because of bad body if at all.

owais commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+	if r.Time == nil {
+		return nil
+	}
+	ts := time.Unix(0, int64(float64(time.Second)**r.Time))
```
This preserves nanoseconds so resultant proto ts has more or less correct value set for the `Nanos` field.

pjanotti replied:
Ah missed that Time was a float...

owais commented on receiver/collectdreceiver/collectd.go:
```diff
@@ -0,0 +1,292 @@
+					cindex = len(dimensions)
+				}
+				piece := dimensions[prev:cindex]
+				tindex := strings.Index(piece, "=")
```
I'm not sure. Decided to change change too much so it's easy to compare this with Gateway in case we discover inconsistent behavior.

owais commented:
Thanks for the detailed review @pjanotti. Updated the PR.

owais commented on Makefile:
```diff
@@ -23,6 +23,7 @@ MISSPELL=misspell -error
 MISSPELL_CORRECTION=misspell -w
 STATICCHECK=staticcheck
 IMPI=impi
+RUN_CONFIG=local/config.yaml
```
I have 30 different config files in the local folder so added it to git. May be we can add a new directory that contains a default config file used by run but I don't see much value in it as it'll mostly by used by devs and each dev will have to use a special config file depending on the component they work on.

pjanotti replied:
Perhaps better as an automated test, but having one that successfully launches the collector seems reasonable. Devs could run it as a sanity check (at least I would). That said it is fine we can add it to a different location.

pjanotti reviewed (approved):
LGTM. Please check the suggestion regarding the README.md

owais commented on exporter/stackdriverexporter/go.sum:
```diff
@@ -22,6 +24,8 @@ github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym/WlBOVXweHU+Q+/VP0lqqI8lqeDx9IjBqo=
 github.com/DataDog/datadog-go v2.2.0+incompatible/go.mod h1:LButxg5PwREeZtORoXG3tL4fMGNddJ+vMq1mwgfaqoQ=
+github.com/Microsoft/ApplicationInsights-Go v0.4.2 h1:HIZoGXMiKNwAtMAgCSSX35j9mP+DjGF9ezfBvxMDLLg=
```
stackdriver? I ran go mod tidy and that updated the go.sum file. I think it's because of the azure exporter.

owais replied:
>  Similar below for azure depending on collectD

Do you mean why stuff related to azure is in this (collectd) PR?

pjanotti replied:
Let's fix that in a separate PR.

owais committed:
Added CollectD JSON receiver

Ported SignalFx CollectD receiver from SignalFX Gateway to OpenTelemetry Collector

Ported from: https://github.com/signalfx/gateway/tree/master/protocol/collectd

owais commented on receiver/collectdreceiver/README.md:
```diff
@@ -0,0 +1,5 @@
+
+This receiver can receive data exported by the CollectD's `write_http` plugin. Only JSON format is supported. Authentication is not supported but support can be added later if needed.
+
+This receiver was donated by SignalFx and ported from SignalFx's Gateway (https://github.com/signalfx/gateway/tree/master/protocol/collectd). As a result, this receiver has supports some additional faetures that are technically not compatible with stock CollectD's write_http plugin. That said, in practice such incompatibilities should never surface. For example, this receiver supports extracting  ...
```
Fixed.
